### PR TITLE
Minor enhancements

### PR DESCRIPTION
**Pull request by mcw0**
The ssh_process() should be closed with except failure
Banner option to not show message about the remote system

# Pwntools Pull Request

Thanks for contributing to Pwntools!  Take a moment to look at [`CONTRIBUTING.md`][contributing] to make sure you're familiar with Pwntools development.

Please provide a high-level explanation of what this pull request is for.

## Testing

Pull Requests that introduce new code should try to add doctests for that code.  See [`TESTING.md`][testing] for more information.

## Target Branch

Depending on what the PR is for, it needs to target a different branch.

You can always [change the branch][change] after you create the PR if it's against the wrong branch.

| Branch   | Type of PR                                                       |
| -------- | ---------------------------------------------------------------- |
| `dev`    | New features, and enhancements
| `dev`    | Documentation fixes and new tests
| `stable` | Bug fixes that affect the current `stable` branch
| `beta`   | Bug fixes that affect the current `beta` branch, but not `stable`
| `dev`    | Bug fixes for code that has never been released

[contributing]: https://github.com/Gallopsled/pwntools/blob/dev/CONTRIBUTING.md
[testing]: https://github.com/Gallopsled/pwntools/blob/dev/TESTING.md
[change]: https://github.com/blog/2224-change-the-base-branch-of-a-pull-request

## Changelog

After creating your Pull Request, please add and push a commit that updates the changelog for the appropriate branch.  
You can look at the existing changelog for examples of how to do this.

### CODE REVIEW

File: pwnlib/tubes/sock.py [hunk 1]
```diff
@@ -112,6 +112,10 @@ def can_recv_raw(self, timeout):
             self.recv_raw(1, socket.MSG_PEEK)
         except EOFError:
             return False
+        # SSLSocket and Paramiko recv() do not support any flags
+        except (ValueError, TypeError):
+            # select() said there is data, so pass
+            pass
```
Comment (heapcrash) [inline pwnlib/tubes/sock.py:118]: This should be in a different pull request

File: pwnlib/tubes/ssh.py [hunk 2]
```diff
@@ -553,7 +553,7 @@ class ssh(Timeout, Logger):
 
     def __init__(self, user=None, host=None, port=22, password=None, key=None,
                  keyfile=None, proxy_command=None, proxy_sock=None,
-                 level=None, cache=True, ssh_agent=False, *a, **kw):
+                 level=None, cache=True, ssh_agent=False, banner=True, *a, **kw):
```
Comment (heapcrash) [inline pwnlib/tubes/ssh.py:556]: I would prefer the name is `checksec` vs `banner`

File: pwnlib/tubes/ssh.py [hunk 3]
```diff
@@ -1090,6 +1092,8 @@ def is_exe(path):
                 h.failure("Process creation failed")
                 self.warn_once('Could not find a Python interpreter on %s\n' % self.host \
                                + "Use ssh.run() instead of ssh.process()")
+                if python.connected():
+                    python.close()
```
Comment (heapcrash) [inline pwnlib/tubes/ssh.py:1096]: We can just call `python.close()` without checking `connected`

**Review by heapcrash — CHANGES_REQUESTED** [review]

**Comment by heapcrash** [comment]
This also needs to update CHANGELOG.md to add info on this pull request

**Comment by heapcrash** [comment]
CI passes, but I think I want to change `context.ssh_session` instead to store it on the `SSHPath` class itself.

`SSHPath.session = ...`

vs

`context.ssh_session = ...`

Since `context.ssh_session` isn't used anywhere else, and might be misleading (e.g. how does it interact with `process()`).

File: pwnlib/tubes/sock.py (hunk 1 above)
Comment (mcw0) [reply pwnlib/tubes/sock.py:118]: Ok, sure, please tell me how you want me to proceed so I don't create greater mess.

**Comment by heapcrash** [comment]
Whoops the above comment was for another PR

File: pwnlib/tubes/ssh.py (hunk 2 above)
Comment (mcw0) [reply pwnlib/tubes/ssh.py:556]: My first approach as well, but changed my mind when I was looking at the 'checksec()'  function itself, and saw 'banner' as argument. Maybe should do a cleanup too, as it seems to be an artifact of some kind?

File: pwnlib/tubes/ssh.py (hunk 3 above)
Comment (mcw0) [reply pwnlib/tubes/ssh.py:1096]: Cool, will do. Even my experience says it's better to do a check before just close()

**Comment by mcw0** [comment]
> This also needs to update CHANGELOG.md to add info on this pull request

Will do. Thx.

**Comment by mcw0** [comment]
> Whoops the above comment was for another PR

No problems :)

**Comment by mcw0** [comment]
Another thing, just to get some insight in pwntools naming scheme (or pse point me to right reading). 

I'm currently looking into system() -> ssh_channel() for running custom python script, to where I must set 'env' for even finding python in the $PATH.  But the 'msg' (msg = 'Opening new channel: %r' % (process or 'shell')) are quite 'debug informative' of my custom python script (I.e. gives a printout of whole script), and the whole session while executing doesn't  look very esthetic. I have some ideas how to suppress the level of output with the argument, you don't by any chance have suggestion for such naming? (maybe even how to suppress the printout of the script)

File: pwnlib/tubes/ssh.py (hunk 2 above)
Comment (heapcrash) [reply pwnlib/tubes/ssh.py:556]: `checksec` shouldn't collide with `self.checksec`,  it should be fine

File: pwnlib/tubes/ssh.py (hunk 3 above)
Comment (heapcrash) [reply pwnlib/tubes/ssh.py:1096]: I'm pretty sure you can call `io.close(); io.close(); io.close();` as many times as you want and nothing bad happens

File: pwnlib/tubes/sock.py (hunk 1 above)
Comment (heapcrash) [reply pwnlib/tubes/sock.py:118]: Can you add a doctest for this case?

**Comment by mcw0** [comment]
You are free to do whatever you like with this, I've made a wrapper around 'can_recv()' w/ timeout=0.010 (stressing that, but seems stable), custom 'exec_command()' that hooking on your 'transport()'.

btw, you might would like to consider to turn off hardcoded zlib SSH compression (or make it as option) , as paramiko has that default off in their 'connect()' for the reason 'By default, compression is off since it negatively affects interactive sessions.'

Have a nice day
